File: src/pann/Core/Neuron.cpp
//Neuron.cpp

#include "Neuron.h"

using namespace std;
using namespace boost;

namespace pann
{
    void
    Neuron::addInConnection(NeuronPtr _to, WeightPtr _weight)
    {
        links_in.push_back( Link(_to, _weight) );
    } //addInConnection

    void
    Neuron::addOutConnection(NeuronPtr _to, WeightPtr _weight)
    {
        links_out.push_back( Link(_to, _weight) );
    } //addOutConnection

    void
    Neuron::delInConnection(NeuronPtr _to)
    {
        //TODO: Fix this shit. Or not?
        struct comparator
        {
            static bool comp(NeuronPtr _to, const Link& _l)
            {
                return _l.getTo() == _to;
            }
        };

        links_in.remove_if(bind(comparator::comp, _to, _1));
    } //delInConnection

    void
    Neuron::delOutConnection(NeuronPtr _to)
    {
        struct comparator
        {
            static bool comp(NeuronPtr _to, const Link& _l)
            {
                return _l.getTo() == _to;
            }
        };

        links_out.remove_if(bind(comparator::comp, _to, _1));
    } //delOutConnection
/*
    void
    PyramidalNeuron::fire()
    {
        if(bias)
            receptiveField += bias->getValue();

        BOOST_FOREACH( const Link& link, links_in )
            receptiveField += link.getTo()->getOutput() * link.getWeight()->getValue();

        activationValue = 0;
        if(activationFunction)
            activationValue += activationFunction->f(receptiveField);

        receptiveField = 0;
    } //fire

    void
    PyramidalNeuron::learn()
    {
    } //learn
*/
}; //pann

File: trunk/src/pypann/Core/Neuron.py
# -*- encoding: utf-8

"""
Neuron
"""

from Weight import Weight
from Link import Link

class Neuron:
    def __init__(self):
        self._links_in = []
        self._links_out = []
        self.input = 0
        self.output = 0

    def run(self):
        pass

class PyramidalNeuron:
    def __init__(self, tf):
        self.Neuron.__init__()
        self.tf = tf

    def run():
        for link in _links_in:
            if self.tf:
                self.input += link.get_to().output * link.get_weight().get_value()
                self.output = self.tf.f(self.input)

#af - раз нейрон содержит Runner, то пусть он и занимается всякими ф-ями авктивации. Пусть хранит ее в нейроне как аттрибут
#bias - Сделать Link, который меняет поведение. Может содержать Weight, а может просто ф-ю. Bias можно реализовать через ф-ю Link-а

"""
Testing
"""
def test_Neuron():
    print "Testing Neuron..."
    def pyramidal_runner(n):
        for l in n._links_in:
            


if __name__ == "__main__":
    test_Neuron()
